app/models.py

```diff
@@ -420,6 +429,26 @@ class CanonicalFactsSchema(MarshmallowSchema):
         fields.Str(validate=marshmallow_validate.Length(min=1, max=59)), validate=marshmallow_validate.Length(min=1)
     )
     external_id = fields.Str(validate=marshmallow_validate.Length(min=1, max=500))
+    provider_id = fields.Str(validate=marshmallow_validate.Length(min=1, max=500))
+    provider_type = fields.Str(validate=marshmallow_validate.Length(min=1, max=50))
+
+    @validates_schema
+    def validate_provider(self, data, **kwargs):
+        provider_type = data.get("provider_type")
+        provider_id = data.get("provider_id")
+
+        if (provider_type and not provider_id) or (provider_id and not provider_type):
+            raise MarshmallowValidationError("provider_type and provider_id are both required.")
```

Comment:
This logic can be simplified as:
```
if provider_type and provider_id:
    ... ...
    ... ...
```
right?

Reply:
Yep, classic logic refactor. Good call!

Reply:
A NAND would fire when both `provider_type` and `provider_id` are present. We want the check to function as an XOR. If they were booleans a simple != would do the trick, but here it would do the wrong comparison.
we could maybe do
```suggestion
        if (provider_type is None) != (provider_id is None):
            raise MarshmallowValidationError("provider_type and provider_id are both required.")
```
I think that might be less readable though

Reply:
Actually, with how the code is currently, if `provider_type = None` and `provider_id = None`, it won't raise a `MarshmallowValidationError` either. That should be a bug, right? :bug:

Reply:
Nope, I was mistaken earlier @tahmidefaz; it's valid if both are provided, or if neither are provided. We effectively need like an xor existence check on this thing

Reply:
Since it's an XOR comparison, I think the current code readability here is good enough. :+1: